### PR TITLE
Reorg crates

### PR DESCRIPTION
**Pull request by koushiro**
~~waiting for #192 to be merged firstly.~~

* reorg the crates: 
  * move `ReadOnlyDB` trait into `substrate-archive-backend`
  * move `types`, `models` and `util` modules into `substrate-archive`
  * merge `block_exec` and `tasks`, `sql_block_builder` and `models` 
  * remove `substrate-archive-common`
* ~~improve error handling (#194)~~

### CODE REVIEW

**Comment by insipx** [comment]
Why get rid of `common`?

Ideally I would like to merge this after #134 , since it would be a PITA to merge all the reorg changes into another big PR

**Comment by koushiro** [comment]
> Why get rid of `common`?
> 
> Ideally I would like to merge after #134 , since it would be a PITA to merge all the reorg changes into another big PR

IMO, `substrate-archive-common` is a boring crate, it just extracts some **non-public** parts from the two crates of `substrate-archive-backend` and `subtrate-archive` (like `ReadOnlyDB` trait used by backend, `types`, `models` and `logger` used by archive), and also messy error types.

I don't mind one more rebase modification after merging #134

**Comment by koushiro** [comment]
@insipx PTAL

**Review by insipx — APPROVED** [review]

**Comment by insipx** [comment]
Overall good changes, condenses code where it was 'hidden' otherwise

**Comment by koushiro** [comment]
@dvdplm @niklasad1 PTAL

File: substrate-archive/src/wasm_tracing.rs [hunk 1]
```diff
@@ -164,6 +164,7 @@ impl TraceHandler {
 
 	// we need this because we don't know the values until after tracing has been executed
 	/// Returns true if a span is part of an enabled Target. Checks WASM in addition to the spans target.
+	#[allow(clippy::suspicious_operation_groupings)]
```
Comment (dvdplm) [inline substrate-archive/src/wasm_tracing.rs:167]: Is this about silencing the complaint on line 175 (suggestion: "use t.level")? If so that seems to be a clippy bug and not something worth silencing like this imo.

File: substrate-archive/src/archive.rs [hunk 2]
```diff
@@ -29,13 +29,12 @@ use sp_runtime::{
...
 
-use substrate_archive_backend::{runtime_api, ReadOnlyBackend, RuntimeConfig, TArchiveClient};
-use substrate_archive_common::{util, ReadOnlyDB};
+use substrate_archive_backend::{runtime_api, ReadOnlyBackend, ReadOnlyDB, RuntimeConfig, TArchiveClient};
 
 use crate::{
 	actors::{System, SystemConfig},
 	error::Result,
-	migrations,
+	migrations, util,
```
Comment (dvdplm) [inline substrate-archive/src/archive.rs:37]:
```suggestion
	migrations, 
	util,
```

File: substrate-archive/src/logger.rs [hunk 3]
```diff
@@ -1,39 +1,18 @@
-// Copyright 2017-2021 Parity Technologies (UK) Ltd.
```
Comment (dvdplm) [inline substrate-archive/src/logger.rs:1]: All files need license headers. Please revert.

File: substrate-archive/src/logger.rs [hunk 4]
```diff
@@ -1,39 +1,18 @@
...
-
-// substrate-archive is distributed in the hope that it will be useful,
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-// GNU General Public License for more details.
-
-// You should have received a copy of the GNU General Public License
-// along with substrate-archive.  If not, see <http://www.gnu.org/licenses/>.
-
-//! logging and general utilities
```
Comment (dvdplm) [inline substrate-archive/src/logger.rs:17]: All modules should have module documentation; if anything improve and extend, but please do not remove any existing docs.

**Review by dvdplm — CHANGES_REQUESTED** [review]
Not sure I see the value of doing this. It's @insipx's call but I'm not convinced.

File: substrate-archive/src/archive.rs (hunk 2 above)
Comment (koushiro) [reply substrate-archive/src/archive.rs:37]: it's result of command `cargo fmt`

File: substrate-archive/src/wasm_tracing.rs (hunk 1 above)
Comment (koushiro) [reply substrate-archive/src/wasm_tracing.rs:167]: Yes, it seems to be a clippy bug (rust-lang/rust-clippy#6722) of rust 1.50, which needs to be fixed by the upstream

Comment (koushiro) [reply substrate-archive/src/wasm_tracing.rs:167]: @dvdplm PTAL

**Comment by koushiro** [comment]
I will improve the running configuration based on this PR, so hope the PR can be reviewed and merged ASAP @insipx @dvdplm

**Review by dvdplm — APPROVED** [review]

File: substrate-archive-backend/src/database.rs [hunk 5]
```diff
@@ -23,7 +23,24 @@ use kvdb::KeyValueDB;
...
+
+// Archive specific K/V database reader implementation
+pub trait ReadOnlyDB: Send + Sync {
+	/// Read key/value pairs from the database
+	fn get(&self, col: u32, key: &[u8]) -> Option<Vec<u8>>;
+	/// Iterate over all blocks in the database
+	fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = KeyValuePair> + 'a>;
+	/// Catch up with the latest information added to the database
+	fn catch_up_with_primary(&self) -> io::Result<()>;
+	// Open database as read-only
```
Comment (niklasad1) [inline substrate-archive-backend/src/database.rs:39]:
```suggestion
	/// Open database as read-only
```

File: substrate-archive/src/error.rs [hunk 6]
```diff
@@ -13,18 +13,21 @@
...
-use std::{env, fmt, io};
+use std::{env, fmt, io, num};
 use thiserror::Error;
 
 pub type Result<T, E = ArchiveError> = std::result::Result<T, E>;
 
 /// Substrate Archive Error Enum
 #[derive(Debug, Error)]
 pub enum ArchiveError {
+	// Rust std error
```
Comment (niklasad1) [inline substrate-archive/src/error.rs:24]: I fine with removing the comment or fixing the comment

```suggestion
	// Rust std io error
```

**Comment by insipx** [comment]
I think this PR is good, and I greatly appreciate your continued contributions.

I think the reason for the delay of this PR is that many of the included changes are stylistic/aesthetic/organizational. While often they make code more readable, substrate-archives codebase is nonetheless under flux and will probably change significantly in the future anyway. So it's sort of sisyphean to try and make the structure perfect now, when there are things like type-decoding/substrate runtime changes/paritydb/pluggable backends in the pipeline that will change the structure anyways. These features also move the project forward toward it's end-goal rather than stagnating on organization. I know it can be annoying to work in a codebase where things aren't where they 'should' be, or where it seems inefficient, but I don't think now is the right time to focus on such changes.


TL;DR I think we should stick with PR's that move the project forward toward it's goal of 'Generic Indexer pluggable for all Substrate Chains' rather than stagnate on organizing the codebase 'correctly'.

In that vein I think PR #227 (I still have to read through all of it, though) is in a better direction, those changes make it clearer to the end user, make config options easier to use and add features which were lacking before (and close two open issues!).


I'll approve this PR but will be more weary towards more stylistic changes in the future

**Comment by koushiro** [comment]
> TL;DR I think we should stick with PR's that move the project forward toward it's goal of 'Generic Indexer pluggable for all Substrate Chains' rather than stagnate on organizing the codebase 'correctly'.

@insipx  Thank you for your detailed reply, I totally agree with it, this will also be what I expected to see
